specs/agents/otel-distribution.md

```diff
@@ -0,0 +1,64 @@
+
+**Elastic OpenTelemetry Distribution**: this is an OpenTelemetry distribution provided by Elastic that is derived from the 
+_Vanilla OpenTelemetry Distribution_.
+
+## General guidelines
+
+These statements are guiding principles of the Elastic OpenTelemetry distributions, they should be considered more as advice than strict rules.
+
+Elastic OpenTelemetry distribution SHOULD ideally:
+- behave as drop-in replacements of their upstream counterparts
+- provide a simple setup and favor onboarding experience (aka "things should work by default").
+- avoid capturing potentially confusing data (see [system metrics](#system-metrics) example below).
+
+## Configuration
+
+Elastic OpenTelemetry distributions MAY override the default configuration.
+When doing so, user-configuration should remain consistent with vanilla distribution:
+- explicit user configuration SHOULD remain effective
+- overriden default configuration MUST have the ability to be restored to upstream default
+
+Elastic specific configuration items MUST be prefixed with `ELASTIC_OTEL_`.
+For example, the [inferred spans](#inferred-spans) feature is configured with `ELASTIC_OTEL_INFERRED_SPANS_*`.
+
+Elastic and platform specific configuration items must be prefixed with `ELASTIC_OTEL_${platform}_` to be consistent with
+the upstream `OTEL_${platform}_` prefix.
+
+For simplicity the configuration in this specification will use the "environment variable" syntax, some platforms like Java
+might also support other ways to configure.
+
+## Features
+
+### Inferred spans
+
+Supported platforms: [Java](https://github.com/elastic/elastic-otel-java/tree/main/inferred-spans)
+
+Configuration namespace: `ELASTIC_OTEL_INFERRED_SPANS_*`
+
+### System metrics
+
+These metrics are usually captured using the collector running locally but in case where no collector is present, or a centralized
+collector is used then the user might opt in to also collect those.
+
+These metrics are not captured by default in order to prevent duplicated metrics when they are also captured by a collector.
+
+TODO : add link to supported platforms and respective implementations
+TODO : add the configuration option name to enable those
+
+### Cloud resource attributes
+
+Supported platforms: Java
+
+The cloud resource attributes ([semconv](https://opentelemetry.io/docs/specs/semconv/resource/cloud/)) provide equivalent 
+attributes to the [cloud provider metadata](metadata.md#cloud-provider-metadata), which are usually provided
+through a metadata HTTP(s) endpoint accessible from the application.
+
+Elastic OpenTelemetry distributions SHOULD capture those by default for a better onboarding experience.
+Users MUST be able to disable this default to minimize application startup overhead or if those attributes are provided through the collector.
+
+Elastic distribution MUST allow resource providers to be disabled through explicit configuration by supporting `OTEL_RESOURCE_PROVIDERS_${provider}_ENABLED=false`.
```

Comment:
We don't have it in python and I can't find any reference to it on the common environment variables https://opentelemetry.io/docs/specs/otel/configuration/sdk-environment-variables/, do you have any link?

Reply:
The implementation is only in the java "autoconfiguration", so it's only implemented for the java agent.

The actual implementation of those providers comes from the contrib repo, and reusing them as-is in an application will enable them by default (due to how auto-configuration works with the SDK). While the implementation here might be very java-specific, the feature is definitely cross-platform, adding a comment that it's java specific for not could help clarify this.

@trentm I remember there are similar providers in the nodejs contrib repo, do you know if there are ways to disable them through configuration, and if yes what is the configuration option ?

Reply:
(Aside on the java envvar:

I see https://opentelemetry.io/docs/languages/java/automatic/configuration/#enable-resource-providers-that-are-disabled-by-default mentions `otel.resource.providers.aws.enabled` but not the environment variable equivalent.

... oh, you need to hit the "(+)" disclosure to see it, and this is generated by a template at https://github.com/open-telemetry/opentelemetry.io/blob/main/content/en/docs/languages/java/automatic/configuration.md#enable-resource-providers-that-are-disabled-by-default   The string `OTEL_RESOURCE_PROVIDERS_` doesn't appear in code anywhere, so my grepping was failing. :)

---

The OTel JS `@opentelemetry/sdk-node` package supports a `export OTEL_NODE_RESOURCE_DETECTORS="env,host"` envvar to limit to a particular set of resource detectors. Documented here:
https://github.com/open-telemetry/opentelemetry-js/tree/main/experimental/packages/opentelemetry-sdk-node#resourcedetectors
By default this package only includes a few resource detectors -- none of the cloud ones.

The separate `@opentelemetry/auto-instrumentations-node` package -- which provides the closest thing to zero-code instrumentation for Node.js -- also supports `OTEL_NODE_RESOURCE_DETECTORS` and includes many more detectors by default (including the cloud ones). It also includes "all" and "none" explicit values.
Documented here: https://github.com/open-telemetry/opentelemetry-js-contrib/tree/main/metapackages/auto-instrumentations-node/#usage-auto-instrumentation

Reply:
@xrmx for python I've found a few resource providers but I don't know if they can be enabled/disabled through configuration, the examples below are only registered through explicit code:
- [azure (contrib)](https://github.com/open-telemetry/opentelemetry-python-contrib/tree/460fc335836c395db8472ecf464e7ecd94c08925/resource/opentelemetry-resource-detector-azure)
- [aws ec3 (contrib)](https://github.com/open-telemetry/opentelemetry-python-contrib/tree/460fc335836c395db8472ecf464e7ecd94c08925/sdk-extension/opentelemetry-sdk-extension-aws#usage-aws-resource-detectors)

So if you know a way to disable them through configuration it would be worth to add a reference to it here.

Reply:
I've added platform-specific examples in 3cef06fea46167a3856e2f5d4b8b8428eb92a839 here so it's more clear.

This also highlights that there is probably an opportunity to contribute a cross-platform way to do this upstream.

Reply:
In upstream python we have `OTEL_EXPERIMENTAL_RESOURCE_DETECTORS` that behaves like the js one, have to ask why it's experimental, probably a very old semantic convention